File: src/main/java/relicsorter/RelicSorter.java
package relicsorter;

import basemod.BaseMod;
import basemod.interfaces.PreDungeonUpdateSubscriber;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Input;
import com.evacipated.cardcrawl.modthespire.lib.SpireConfig;
import com.evacipated.cardcrawl.modthespire.lib.SpireInitializer;
import com.megacrit.cardcrawl.characters.AbstractPlayer;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import com.megacrit.cardcrawl.helpers.input.InputHelper;
import com.megacrit.cardcrawl.relics.AbstractRelic;

import java.io.IOException;

@SpireInitializer
public class RelicSorter implements PreDungeonUpdateSubscriber{
	
    public static final String NAME = "Relic Sorter";
	public static final String AUTHOR = "Blank The Evil";
	public static final String DESC = "Allows you to send relics to the front of the list by right clicking them.";
	public static BindingEnum binding;
	public static int keyBind;
	public enum BindingEnum {
	    LEFT,
        RIGHT,
        MIDDLE
    }

	public RelicSorter() {
		BaseMod.subscribe(this);
		BaseMod.subscribe(new RelicSorterInit());
        load();
    }

    public static void load(){
	    try {
            SpireConfig config = new SpireConfig("relicSorter", "config");
            config.load();
            String valueE = config.getString("BindingEnum");
            int valueK = config.getInt("KeyBinding");
            if(valueE != null){
                binding = BindingEnum.valueOf(valueE);
            }
            keyBind = valueK;
        }catch(IOException | NumberFormatException e){
	        log("Failed to load settings!");
	        binding = BindingEnum.RIGHT;
	        keyBind = Input.Keys.R;
        }
    }

    public static void save(){
        try {
            SpireConfig config = new SpireConfig("relicSorter", "config");
            config.setString("BindingEnum", binding.toString());
            config.setInt("KeyBinding", keyBind);

            log("Saving keybinding as: " + Input.Keys.toString(keyBind));

            config.save();
        }catch (IOException e){
            log("Failed to save settings!");
        }
    }

    @Override
    public void receivePreDungeonUpdate() {
	    if(AbstractDungeon.player == null || AbstractDungeon.player.relics == null)
	        return;

	    AbstractPlayer player = AbstractDungeon.player;

        boolean clicked = false;
        switch(binding){
            case LEFT:
                clicked = InputHelper.justClickedLeft;
                break;
            case RIGHT:
                clicked = InputHelper.justClickedRight;
                break;
            case MIDDLE:
                clicked = handleMiddleClick();
                break;
        }
        if(!clicked)
            clicked = Gdx.input.isKeyJustPressed(keyBind);

        if(player.relics.size() > 2 && clicked) {
            int selIndex = -1;
            for (int i = player.relics.size() - 1; i >= 2; i--) {
                AbstractRelic relic = player.relics.get(i);
                if (relic.hb.hovered) {
                    selIndex = i;
                    log("Clicked", relic.name, "" + i);
                }
            }
            if(selIndex != -1) {
                float tempCX = player.relics.get(selIndex).currentX;
                float temphX = player.relics.get(selIndex).hb.x;

                moveRelic(player.relics.get(selIndex), player.relics.get(1));

                AbstractRelic temp = player.relics.remove(selIndex);
                player.relics.add(1, temp);


                for (int i = 2; i < selIndex; i++) {
                    AbstractRelic r1 = player.relics.get(i);
                    AbstractRelic r2 = player.relics.get(i + 1);
                    log("Moving", r1.name, "to", r2.name);
                    moveRelic(r1, r2);
                }

                moveRelic(player.relics.get(selIndex), tempCX, temphX);

                switch (binding) {
                    case LEFT:
                        InputHelper.justClickedLeft = false;
                        break;
                    case RIGHT:
                        InputHelper.justClickedRight = false;
                        break;
                    case MIDDLE:
                        break;
                }
            }
        }
    }

    public static void log(String ... items) {
        System.out.print("RelicSorter ");
        for(String item : items) {
            System.out.print(" : " + item);
        }
        System.out.println();
    }

    public static void initialize() {
        log("Version", "1.1.1");
        new RelicSorter();
    }

    private static void moveRelic(AbstractRelic r1, AbstractRelic r2){
        r1.currentX = r2.currentX;
        r1.hb.x = r2.hb.x;
    }

    private static void moveRelic(AbstractRelic r1, float currentX, float hbX){
        r1.currentX = currentX;
        r1.hb.x= hbX;
    }

    private static boolean justPressedMiddleLast = false;

    private static boolean handleMiddleClick(){
        boolean isMiddleDown = Gdx.input.isButtonPressed(Input.Buttons.MIDDLE);
        if(isMiddleDown && !justPressedMiddleLast) {
            justPressedMiddleLast = true;
            return true;
        }

        if(!isMiddleDown && justPressedMiddleLast)
            justPressedMiddleLast = false;

        return false;
    }
	
}

File: src/main/kotlin/com/blanktheevil/relicsorter/RelicSorter.kt
package com.blanktheevil.relicsorter

import basemod.BaseMod
import basemod.interfaces.PreDungeonUpdateSubscriber
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Input
import com.blanktheevil.relicsorter.helpers.RelicSorterHelper
import com.blanktheevil.relicsorter.models.BindingEnum
import com.blanktheevil.relicsorter.models.Config
import com.evacipated.cardcrawl.modthespire.lib.SpireInitializer
import com.megacrit.cardcrawl.core.Settings
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import com.megacrit.cardcrawl.helpers.input.InputHelper
import com.megacrit.cardcrawl.relics.AbstractRelic
import java.io.IOException
import java.util.*


@Suppress("unused")
@SpireInitializer
class RelicSorter : PreDungeonUpdateSubscriber {
  companion object {

    lateinit var config: Config

    var name: String = ""
    private var version: String = ""
    var modid: String = ""
    var author: String = ""
    var description: String = ""

    @JvmStatic
    fun initialize() {
      loadProjectProperties()
      config = Config.init()
      BaseMod.subscribe(RelicSorterInit())
      BaseMod.subscribe(RelicSorter())
    }

    private fun loadProjectProperties() {
      try {
        with(Properties()) {
          load(RelicSorter::class.java.getResourceAsStream("/META-INF/relicsorter.prop"))
          name = getProperty("name")
          version = getProperty("version")
          modid = getProperty("id")
          author = getProperty("author")
          description = getProperty("description")
        }
      } catch (e: IOException) {
        e.printStackTrace()
      }
    }

    private fun log(vararg items: Any?) {
      if (Settings.isDebug) {
        print("Relic Sorter")
        items.forEach {
          print(" : ${it.toString()}")
        }
        println()
      }
    }
  }

  override fun receivePreDungeonUpdate() {
    if (AbstractDungeon.player == null || AbstractDungeon.player.relics == null) return
    with(AbstractDungeon.player) {
      var clicked = when (config.binding) {
        BindingEnum.LEFT -> InputHelper.justClickedLeft
        BindingEnum.RIGHT -> InputHelper.justClickedRight
        BindingEnum.MIDDLE -> handleMiddleClick()
        BindingEnum.NONE -> false
      }

      if (!clicked) clicked = Gdx.input.isKeyJustPressed(config.keyBind)

      if (relics.size > 2 && RelicSorterHelper.canMoveRelics() && clicked) {
        var selIndex: Int? = null

        for (i in relics.size - 1 downTo 2) {
          val relic: AbstractRelic = relics[i]
          if (relic.hb.hovered) {
            selIndex = i
            log("Clicked ${relic.name} $i")
          }
        }

        if (selIndex != null) {
          val tempCurrentX = relics[1].currentX

          for (i in 1 until selIndex) {
            val relicA = relics[i]
            val relicB = relics[i + 1]
            log("Moving ${relicA.name} to ${relicB.name}")
            relicA.moveTo(relicB)
          }

          relics[selIndex].moveTo(tempCurrentX)
          relics.add(1, relics.removeAt(selIndex))

          when (config.binding) {
            BindingEnum.LEFT -> InputHelper.justClickedLeft = false
            BindingEnum.RIGHT -> InputHelper.justClickedRight = false
            else -> { /* do nothing */ }
          }
        }
      }
    }
  }

  private var justPressedMiddleLast = false

  private fun handleMiddleClick(): Boolean {
    val isMiddleDown = Gdx.input.isButtonPressed(Input.Buttons.MIDDLE)
    if (isMiddleDown && !justPressedMiddleLast) {
      justPressedMiddleLast = true
      return true
    }

    if(!isMiddleDown && justPressedMiddleLast) {
      justPressedMiddleLast = false
    }

    return false
  }
}